std/machines/memory16.asm

```diff
@@ -51,26 +55,38 @@ machine Memory16(byte2: Byte2) with
     (1 - m_is_write') * (1 - m_change) * (m_value1' - m_value1) = 0;
     (1 - m_is_write') * (1 - m_change) * (m_value2' - m_value2) = 0;
 
-    col witness m_diff_lower;
-    col witness m_diff_upper;
-
     col fixed FIRST = [1] + [0]*;
     let LAST = FIRST';
 
-    link => byte2.check(m_diff_lower);
-    link => byte2.check(m_diff_upper);
-
     std::utils::force_bool(m_change);
 
     // if change is zero, addr has to stay the same.
-    (m_addr' - m_addr) * (1 - m_change) = 0;
+    (m_addr_low' - m_addr_low) * (1 - m_change) = 0;
+    (m_addr_high' - m_addr_high) * (1 - m_change) = 0;
 
-    // Except for the last row, if change is 1, then addr has to increase,
+    // Except for the last row, if m_change is 1, then addr has to increase,
     // if it is zero, step has to increase.
-    // `m_diff_upper * 2**16 + m_diff_lower` has to be equal to the difference **minus one**.
-    // Since we know that both addr and step can only be 32-Bit, this enforces that
-    // the values are strictly increasing.
-    // TODO check this
-    col diff = (m_change * (m_addr' - m_addr)) + (1 - m_change) * (m_step' - m_step);
-    (1 - LAST) * (diff - 1 - m_diff_upper * 2**16 - m_diff_lower) = 0;
+    // The diff has to be equal to the difference **minus one**.
+
+    col witness m_diff_lower;
+    col witness m_diff_upper;
+
+    link => byte2.check(m_diff_lower);
+
+    // When comparing time steps, a 24-Bit diff is sufficient assuming a maximum step
+    // of 2**24.
```

Comment:
I'm not sure we should make this assumption. Is this already taking into account that step grows 4 by row?

Reply:
Increasing by 4 in every step just means that it'll reach that maximum faster, after $2^{22}$ rows ;)

As mentioned, we can increase the range constraint of `m_diff_upper` to 12 bits. That'll give us a maximum diff of $2^{28}$, so $2^{26}$ rows.

Reply:
Yea I think we should do that. If we allow for main machines of up to 24, main regs will get to 26 in some cases, which I think we should support if it's not too much extra work.

Reply:
Done